File: app/src/main/java/com/example/se_proj/WalkInRegistrationActivity.java
package com.example.se_proj;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.example.se_proj.databinding.ActivityWalkInRegistrationBinding;
import com.example.se_proj.models.VisitorRequest;
import com.example.se_proj.rules.RequestStatus;
import com.example.se_proj.rules.RequestValidationUtils;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.ListenerRegistration;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Captures guard-initiated walk-in visitor requests and waits for host approval in real time.
 *
 * <p>Design note: event-driven workflow using a Firestore document listener to update UI as the
 * request status transitions from {@code pending_adhoc} to {@code approved}/{@code denied}.</p>
 *
 * <p>Outstanding issues: listener lifecycle is tied to this screen only; if the guard navigates
 * away, approval updates are missed until the request is searched again.</p>
 */
public class WalkInRegistrationActivity extends AppCompatActivity {

    private ActivityWalkInRegistrationBinding binding;
    private final FirebaseFirestore db = FirebaseFirestore.getInstance();
    private final FirebaseAuth auth = FirebaseAuth.getInstance();
    private ListenerRegistration listenerRegistration;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityWalkInRegistrationBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        binding.toolbar.setNavigationOnClickListener(v -> finishWithoutAnimation());
        binding.toolbar.setOnMenuItemClickListener(item -> {
            if (item.getItemId() == R.id.action_logout) {
                FirebaseAuth.getInstance().signOut();
                Intent intent = new Intent(this, LoginActivity.class);
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                startActivity(intent);
                overridePendingTransition(0, 0);
                finishWithoutAnimation();
                return true;
            }
            return false;
        });

        binding.btnSubmitWalkIn.setOnClickListener(v -> submitWalkInRequest());
        setupBottomNavigation();
    }

    private void setupBottomNavigation() {
        binding.bottomNavigation.setSelectedItemId(R.id.nav_adhoc);
        binding.bottomNavigation.setOnItemSelectedListener(item -> {
            int id = item.getItemId();
            if (id == R.id.nav_adhoc) {
                return true;
            } else if (id == R.id.nav_home) {
                startGuardPortalActivity(new Intent(this, GuardDashboardActivity.class), true);
                return true;
            } else if (id == R.id.nav_logs) {
                Intent intent = new Intent(this, AdminAuditActivity.class);
                intent.putExtra(AdminAuditActivity.EXTRA_SHOW_GUARD_NAV, true);
                startGuardPortalActivity(intent, true);
                return true;
            } else if (id == R.id.nav_settings) {
                Toast.makeText(this, "Settings coming soon", Toast.LENGTH_SHORT).show();
                return true;
            }
            return false;
        });
    }

    private void startGuardPortalActivity(Intent intent, boolean finishCurrent) {
        startActivity(intent);
        overridePendingTransition(0, 0);
        if (finishCurrent) {
            finishWithoutAnimation();
        }
    }

    private void submitWalkInRequest() {
        String hostId = binding.etHostId.getText().toString().trim();
        String name = binding.etGuestName.getText().toString().trim();
        String cnic = RequestValidationUtils.normalizeCnic(binding.etCnic.getText().toString());
        String purpose = binding.etPurpose.getText().toString().trim();

        RequestValidationUtils.ValidationResult validation =
                RequestValidationUtils.validateWalkInRequest(hostId, name, cnic, purpose);
        if (!validation.isValid()) {
            Toast.makeText(this, validation.getMessage(), Toast.LENGTH_SHORT).show();
            return;
        }

        String uid = auth.getCurrentUser() != null ? auth.getCurrentUser().getUid() : null;
        if (uid == null || uid.isEmpty()) {
            Toast.makeText(this, "Please log in again", Toast.LENGTH_LONG).show();
            return;
        }

        String currentTime = new SimpleDateFormat("HH:mm", Locale.getDefault()).format(new Date());
        String currentDate = new SimpleDateFormat("dd/MM/yyyy", Locale.getDefault()).format(new Date());

        VisitorRequest request = new VisitorRequest(
                "", name, cnic, purpose, currentDate, currentTime, "23:59",
                hostId, uid, "faculty", RequestStatus.PENDING_ADHOC, null, null, false
        );

        binding.btnSubmitWalkIn.setEnabled(false);

        db.collection("visitor_requests")
                .add(request)
                .addOnSuccessListener(documentReference -> {
                    Toast.makeText(this, "Request sent to Faculty for approval", Toast.LENGTH_SHORT).show();
                    listenForApproval(documentReference.getId());
                })
                .addOnFailureListener(e -> {
                    binding.btnSubmitWalkIn.setEnabled(true);
                    Log.e("WalkIn", "Failed to send request", e);
                    Toast.makeText(this, "Failed to send request: " + e.getMessage(), Toast.LENGTH_SHORT).show();
                });
    }

    private void listenForApproval(String requestId) {
        binding.tvApprovalStatus.setVisibility(View.VISIBLE);
        binding.tvApprovalStatus.setText("Waiting for Faculty Approval...");

        listenerRegistration = db.collection("visitor_requests").document(requestId)
                .addSnapshotListener((snapshot, e) -> {
                    if (e != null) {
                        Log.w("FirestoreListener", "Listen failed.", e);
                        return;
                    }

                    if (snapshot != null && snapshot.exists()) {
                        String status = snapshot.getString("status");
                        if ("approved".equals(status)) {
                            binding.tvApprovalStatus.setText("APPROVED! You can now check them in.");
                            binding.tvApprovalStatus.setTextColor(
                                    getResources().getColor(android.R.color.holo_green_dark, getTheme()));
                            Toast.makeText(this, "Request Approved!", Toast.LENGTH_LONG).show();
                        } else if ("denied".equals(status)) {
                            binding.tvApprovalStatus.setText("DENIED by Faculty.");
                            binding.tvApprovalStatus.setTextColor(
                                    getResources().getColor(android.R.color.holo_red_dark, getTheme()));
                            binding.btnSubmitWalkIn.setEnabled(true);
                        }
                    }
                });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (listenerRegistration != null) {
            listenerRegistration.remove();
        }
    }

    private void finishWithoutAnimation() {
        finish();
        overridePendingTransition(0, 0);
    }
}

File: app/src/main/java/com/example/se_proj/WalkInRegistrationActivity.kt
package com.example.se_proj

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.se_proj.databinding.ActivityWalkInRegistrationBinding
import com.example.se_proj.models.VisitorRequest
import com.example.se_proj.rules.RequestStatus
import com.example.se_proj.rules.RequestValidationUtils
import com.google.firebase.Firebase
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.firestore
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class WalkInRegistrationActivity : AppCompatActivity() {

    private lateinit var binding: ActivityWalkInRegistrationBinding
    private val db = Firebase.firestore
    private val auth = FirebaseAuth.getInstance()
    private var listenerRegistration: ListenerRegistration? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityWalkInRegistrationBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.toolbar.setNavigationOnClickListener { finish() }
        binding.toolbar.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.action_logout -> {
                    FirebaseAuth.getInstance().signOut()
                    val intent = Intent(this, LoginActivity::class.java)
                    intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
                    startActivity(intent)
                    finish()
                    true
                }
                else -> false
            }
        }

        binding.btnSubmitWalkIn.setOnClickListener { submitWalkInRequest() }
    }

    private fun submitWalkInRequest() {
        val hostId = binding.etHostId.text.toString().trim()
        val name = binding.etGuestName.text.toString().trim()
        val cnic = RequestValidationUtils.normalizeCnic(binding.etCnic.text.toString())
        val purpose = binding.etPurpose.text.toString().trim()

        val validation = RequestValidationUtils.validateWalkInRequest(hostId, name, cnic, purpose)
        if (!validation.isValid) {
            Toast.makeText(this, validation.message, Toast.LENGTH_SHORT).show()
            return
        }

        val uid = auth.currentUser?.uid
        if (uid.isNullOrEmpty()) {
            Toast.makeText(this, "Please log in again", Toast.LENGTH_LONG).show()
            return
        }

        val currentTime = SimpleDateFormat("HH:mm", Locale.getDefault()).format(Date())
        val currentDate = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault()).format(Date())

        val request = VisitorRequest(
            guestName = name,
            guestCNIC = cnic,
            purpose = purpose,
            visitDate = currentDate,
            startTime = currentTime,
            endTime = "23:59",
            hostId = hostId,
            creatorId = uid,
            hostType = "faculty",
            status = RequestStatus.PENDING_ADHOC,
            onCampus = false
        )

        binding.btnSubmitWalkIn.isEnabled = false

        db.collection("visitor_requests")
            .whereEqualTo("creatorId", uid)
            .whereEqualTo("guestCNIC", cnic)
            .whereEqualTo("visitDate", currentDate)
            .get()
            .addOnSuccessListener { documents ->
                val hasDuplicate = documents.toObjects(VisitorRequest::class.java).any {
                    RequestValidationUtils.matchesDuplicateCandidate(it, hostId, cnic, currentDate)
                }
                if (hasDuplicate) {
                    binding.btnSubmitWalkIn.isEnabled = true
                    Toast.makeText(this, "A walk-in request already exists for this guest today", Toast.LENGTH_LONG).show()
                    return@addOnSuccessListener
                }

                db.collection("visitor_requests")
                    .add(request)
                    .addOnSuccessListener { documentReference ->
                        Toast.makeText(this, "Request sent to Faculty for approval", Toast.LENGTH_SHORT).show()
                        listenForApproval(documentReference.id)
                    }
                    .addOnFailureListener { e ->
                        binding.btnSubmitWalkIn.isEnabled = true
                        Log.e("WalkIn", "Failed to send request", e)
                        Toast.makeText(this, "Failed to send request: ${e.message}", Toast.LENGTH_SHORT).show()
                    }
            }
            .addOnFailureListener { e ->
                binding.btnSubmitWalkIn.isEnabled = true
                Log.e("WalkIn", "Failed to validate request", e)
                Toast.makeText(this, "Failed to validate request: ${e.message}", Toast.LENGTH_SHORT).show()
            }
    }

    private fun listenForApproval(requestId: String) {
        binding.tvApprovalStatus.visibility = View.VISIBLE
        binding.tvApprovalStatus.text = "Waiting for Faculty Approval..."

        listenerRegistration = db.collection("visitor_requests").document(requestId)
            .addSnapshotListener { snapshot, e ->
                if (e != null) {
                    Log.w("FirestoreListener", "Listen failed.", e)
                    return@addSnapshotListener
                }

                if (snapshot != null && snapshot.exists()) {
                    val status = snapshot.getString("status")
                    if (status == "approved") {
                        binding.tvApprovalStatus.text = "APPROVED! You can now check them in."
                        binding.tvApprovalStatus.setTextColor(resources.getColor(android.R.color.holo_green_dark, theme))
                        Toast.makeText(this, "Request Approved!", Toast.LENGTH_LONG).show()
                    } else if (status == "denied") {
                        binding.tvApprovalStatus.text = "DENIED by Faculty."
                        binding.tvApprovalStatus.setTextColor(resources.getColor(android.R.color.holo_red_dark, theme))
                        binding.btnSubmitWalkIn.isEnabled = true
                    }
                }
            }
    }

    override fun onDestroy() {
        super.onDestroy()
        listenerRegistration?.remove()
    }
}
